Enable traversal of aliased types

andrewvc opened the pull request:
Currently this library will only traverse `map[string]interface{}` and `[]interface{}` types correctly. This is a problem for us in beats because we usually use `common.MapStr` which uses `map[string]interface{}` as the underlying type. When we switched to this library `Strict()` silently broke in that it would no longer detect extra fields for `common.MapStr`. 

This patch fixes this by using reflection for all underlying operations. Sadly, this is the most efficient way to work here since go lacks generics. Since this is a testing library, the perf hit here is negligible. It's still very fast.

andrewkroh commented on core.go:
```diff
@@ -96,35 +96,46 @@ func Strict(laxValidator validator.Validator) validator.Validator {
 	}
 }
 
+var strType = reflect.TypeOf("")
```
Are these unused?

andrewvc replied:
yes, apologies, I missed that, that was left over from an experiment.

andrewkroh commented on core.go:
```diff
@@ -96,35 +96,46 @@ func Strict(laxValidator validator.Validator) validator.Validator {
+
+		default:
+			// By default we just check reflection equality
+			if inVal.Kind() == reflect.Map {
```
This looks like it can be dropped because it's included as a case in the switch.

urso commented on core.go:
```diff
@@ -96,35 +96,46 @@ func Strict(laxValidator validator.Validator) validator.Validator {
+		case reflect.Map:
+			return compileMap(inVal)
+		case reflect.Slice:
+			return compileSlice(inVal)
```
fyi; reflect.Arary and reflect.Slice are distinguished types :)

urso commented on walk.go:
```diff
@@ -105,8 +108,15 @@ func walkFull(o interface{}, root map[string]interface{}, path llpath.Path, expa
+	}
+
+	for _, kVal := range mVal.MapKeys() {
+		vVal := reflect.ValueOf(mVal.MapIndex(kVal).Interface())
```
MapIndex already returns `reflect.Value`. Why get the interface and call ValueOf again.

andrewvc replied:
That is a bug, good catch.

urso commented on walk.go:
```diff
@@ -105,8 +108,15 @@ func walkFull(o interface{}, root map[string]interface{}, path llpath.Path, expa
+
+	for _, kVal := range mVal.MapKeys() {
+		vVal := reflect.ValueOf(mVal.MapIndex(kVal).Interface())
+		k := kVal.String()
```
what if `kVal.Kind() != reflect.String`?

andrewvc replied:
Good point. I think that this is where things get weird.

I would like the library to handle arbitrary map types in the future, including non-string keys. My thinking is that this gets us partly there. 

In a future version we could internally represent keys as `fmt.Sprintf("%T:%s", kVal.Interface(), value)`.

andrewvc replied:
But to answer your question, the answer for today is that we don't handle that well, but we'd probably do what people expect in most cases.

andrewvc replied:
OK, having messed around with this a bit I think we need a runtime check to validate maps that are not string keyed as scalar values. So, you can still validate them, but you need to do it in one isdef, essentially treating it as a scalar value.

urso commented on walk.go:
```diff
@@ -36,65 +36,68 @@ type walkObserver func(info walkObserverInfo) error
+	// Unpack any wrapped interfaces
+	for oVal.Kind() == reflect.Interface {
+		oVal = reflect.ValueOf(oVal.Interface())
+	}
```
what about pointers? What if you have nested pointers and interfaces?

some helpers from go-ucfg:
```
func chaseValueInterfaces(v reflect.Value) reflect.Value {
	for v.Kind() == reflect.Interface && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseValuePointers(v reflect.Value) reflect.Value {
	for v.Kind() == reflect.Ptr && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseValue(v reflect.Value) reflect.Value {
	for (v.Kind() == reflect.Ptr || v.Kind() == reflect.Interface) && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseTypePointers(t reflect.Type) reflect.Type {
	for t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	return t
}
```

andrewvc replied:
That's a good point (no pun intended).

I think there's an argument that the pointer reflects the actual value. You may actually want to check pointer equality in other words. That said, it really gets thorny here. I feel like interfaces are the right line for this library given that our main goal here is getting to something JSON-ish. 

That said, I may not be seeing a lot of use cases. Does the above line of thinking change still leave you thinking that we should chase pointers? I definitely see how some people might want them.

andrewvc replied:
Thinking about this more, I think you may be right about chasing pointers. I think a goal of this lib should be to approximate JSON-like structures, and in JSON all types are nullable, which means pointers should be able to go anywhere. 

Let me think about this for a bit, but I suspect you're right.

andrewvc replied:
Having thought about it, I think you're right, I added in pointer chasing using `ChaseValue`.

urso commented on walk.go:
```diff
@@ -36,65 +36,68 @@ type walkObserver func(info walkObserverInfo) error
 	if lastPathComponent == nil {
 		// In the case of a slice we can have an empty path
-		if _, ok := o.([]interface{}); ok {
+		if oVal.Kind() == reflect.Interface {
```
I'm a little confused here. Before you did check if it's a slice of interfaces, but now `reflect.Interface` is good enough?

If you 'know' the type you want to check against you can also do this:

```
	var tInterfaceArray = reflect.TypeOf([]interface{}(nil))
    var tMapStringInterface = reflect.TypeOf((map[string]interface{})(nil))
```

These values can be used like:

```
  var val reflect.Type = ...
  if val.Type() == tMapStringInterface {
    ...
  }
```

or 

```
  var val reflect.Type = ...
  if val.Type().ConvertibleTo(tMapStringInterface) {
    m := val.Convert(tMapStringInterface).Interface().(map[string]interface{})
    ...
  }
```

This later trick actually helps in converting aliases of `map[string]interface{}` into `map[string]interface{}`. Then you could proceed without reflection ;)

andrewvc replied:
Thanks for sharing the neat trick! I think that's nifty, but given that we have this reflection based implementation, I think it's nice that it handles arbitrary value types for maps. So, now we can traverse (and as a result give nicer test output) for `map[string]string`, `map[string]int` etc etc.

And, it lays the ground work for better testing `map[int]uint64` and other weirder maps.

WDYT?

andrewvc replied:
I forgot to answer about slices since I got maps in my head. I'm fine to switch slices back to `[]interface{}`, I only switched to reflection for consistency with map traversal.

andrewvc replied:
Also, about the original line, this is a flat out bug from a bulk replace, will push a real fix.

andrewvc replied:
I think the most current code works right and still works correctly with `map[string]string` or whatever arbitrary value type, which is a win, since it gives us richer validation.

andrewvc commented:
@urso I think I mostly addressed your points.

As a next step I think we need runtime checks that all maps are string keyed and better docs around that fact, as well as the pointer / interface chasing behavior for values.
